core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -264,7 +264,11 @@ class LogSegment(val log: FileRecords,
     var lastIndexEntry = 0
     maxTimestampSoFar = RecordBatch.NO_TIMESTAMP
     try {
-      for (batch <- log.batches(maxMessageSize).asScala) {
+      // record batches can grow during compaction (see KAFKA-5316), so we allow some extra space when reading
+      // the batches. Checking the message size is only used as a way to check corruption before checking the CRC
+      val batches = log.batches(maxMessageSize * 2).asScala
```

Comment:
I talked with Jun offline and the suggestion we had is to use the remaining size of the file as the sanity check instead. If we find a message which claims to occupy more space than is possible, then it is obviously corrupt.